larreco/TrackFinder/PMAlgTrackMaker_module.cc

```diff
@@ -0,0 +1,1368 @@
+  static const std::string kNodesName;        // pma nodes
+};
+// -------------------------------------------------------------
+const std::string PMAlgTrackMaker::kKinksName = "kink";
+const std::string PMAlgTrackMaker::kNodesName = "node";
+// -------------------------------------------------------------
+
+PMAlgTrackMaker::PMAlgTrackMaker(fhicl::ParameterSet const & p) :
+	fDetProp(lar::providerFrom<detinfo::DetectorPropertiesService>()),
+	fProjectionMatchingAlg(p.get< fhicl::ParameterSet >("ProjectionMatchingAlg")),
+	fPMAlgVertexing(p.get< fhicl::ParameterSet >("PMAlgVertexing"))
+{
+	this->reconfigure(p);
+
+	produces< std::vector<recob::Track> >();
+	produces< std::vector<recob::SpacePoint> >();
+	produces< std::vector<recob::Vertex> >(); // no instance name for interaction vertices
+	produces< std::vector<recob::Vertex> >(kKinksName); // collection of kinks on tracks
+	produces< std::vector<recob::Vertex> >(kNodesName); // collection of pma nodes
+	produces< std::vector<anab::T0> >();
+
+	produces< art::Assns<recob::Track, recob::Hit> >(); // ****** REMEMBER to remove when FindMany improved ******
+	produces< art::Assns<recob::Track, recob::Hit, recob::TrackHitMeta> >();
+
+	produces< art::Assns<recob::Track, recob::SpacePoint> >();
+	produces< art::Assns<recob::SpacePoint, recob::Hit> >();
+	produces< art::Assns<recob::Vertex, recob::Track> >(); // no instance name for assns of tracks to interaction vertices
+	produces< art::Assns<recob::Track, recob::Vertex> >(kKinksName);  // assns of kinks to tracks
+	produces< art::Assns<recob::Track, anab::T0> >();
+
+	produces< std::vector<recob::PFParticle> >();
+	produces< art::Assns<recob::PFParticle, recob::Cluster> >();
+	produces< art::Assns<recob::PFParticle, recob::Vertex> >();
+	produces< art::Assns<recob::PFParticle, recob::Track> >();
+}
+// ------------------------------------------------------
+
+void PMAlgTrackMaker::beginJob()
+{
+	art::ServiceHandle<art::TFileService> tfs;
+	fTree_trk = tfs->make<TTree>("PMAlgTrackMaker_trk", "tracks overall info");
+	fTree_trk->Branch("fEvNumber", &fEvNumber, "fEvNumber/I");
+	fTree_trk->Branch("fTrkIndex", &fTrkIndex, "fTrkIndex/I");
+	fTree_trk->Branch("fLength", &fLength, "fLength/D");
+	fTree_trk->Branch("fHitsMse", &fHitsMse, "fHitsMse/D");
+	fTree_trk->Branch("fSegAngMean", &fSegAngMean, "fSegAngMean/D");
+	fTree_trk->Branch("fPidTag", &fPidTag, "fPidTag/I");
+}
+
+void PMAlgTrackMaker::reconfigure(fhicl::ParameterSet const& pset)
+{
+	fHitModuleLabel = pset.get< std::string >("HitModuleLabel");
+	fCluModuleLabel = pset.get< std::string >("ClusterModuleLabel");
+	fCluMatchingAlg = pset.get< int >("CluMatchingAlg");
+
+	fTrackingOnlyPdg = pset.get< std::vector<int> >("TrackingOnlyPdg");
+	fTrackingSkipPdg = pset.get< std::vector<int> >("TrackingSkipPdg");
+
+	fMakePFPs = pset.get< bool >("MakePFPs");
+
+	fMinSeedSize1stPass = pset.get< size_t >("MinSeedSize1stPass");
+	fMinSeedSize2ndPass = pset.get< size_t >("MinSeedSize2ndPass");
```

Comment:
It would be a good idea to validate inputs.  There are new features coming in art to assist with this.